### DISCUSSION TITLE
Does qdrant support prefiltering on the fields in the payload before doing the vector embedding match?

### DISCUSSION THREAD

**Discussion by dingusagar** [qdrant/qdrant#322] [Q&A] [answered]
if there are lots of documents, we would want to first filter them on some fields like category etc, and then do vector similarity on a text field, wanted to know if qdrant supports this kind of prefiltering. Or does it do the vector similarity on the entire documents and then filter based on the fields specified in the query?

**Comment by generall** [comment, marked as answer]
Hi @dingusagar, thanks for the great question! The short answer: we do filtering **during** the vector search.
More detailed answer: it depends.

For different search scenario Qdrant query planner prefers different strategies of searching. In case if filter query is restrictive - it became much faster to just retrieve vectors by filtering conditions and then re-score them by similarity.

In more complicated scenario - it perform search using vector index. But unlike other implementations we do not calculate the filtering condition for all the documents in advance - we check conditions dynamically during the traversal of HNSW graph. That allow us to limit the number of condition checks by the order of magnitude comparing with usual pre-filtering approach.

There are more details on how we treat HNSW filtering in the documentation https://qdrant.tech/documentation/indexing/#filtrable-index

P.S. I am going to convert this issue into discussion.

**Reply by dingusagar** [reply]
Thanks for the quick reply. This modified version of HNSW that supports filtering sounds interesting. will check out the paper to learn more.

**Reply by failable** [reply]
Hi, does 'filtering during the vector search' mean that points that satisfy the filter conditions (e.g., "name" == "Tom") and have a high similarity to the query might not be retrieved?

**Reply by Yevgnen** [reply]
Same question.

**Reply by dingusagar** [reply]
@liebkne , @Yevgnen . For requests having a filter condition,  even if the points have high similarity, it won't be returned if the filter condition is not met.

**Reply by Yevgnen** [reply]
Hi, think you misread his question. The question means 'are points satisfying filtering conditions with high similarity score guaranteed to be retrieved under this filtering during the search fashion?'

I've encountered some cases where the data points are not retrieved even though the similarity is extremely high and the filter criteria are met. This seems to happen when there are fewer data points.

That is, from the results, is this filtering supposed to be the same as pre filtering?

**Reply by dingusagar** [reply]
Oh I see, let's wait for @generall to reply regarding this. 
From my experience working with qdrant, I did not face any such issues. I was dealing with collections with 10k to 40k vectors.

**Reply by generall** [reply]
Filtered search is also approximate nearest neighbors. In the same way as regular search doesn't guarantee all results, filtered search doesn't guarantee them either. However, you can control precision with ef and m parameters.

**Reply by failable** [reply]
@generall Thanks for the clarification!
